docs/proposals/20210335-load-balancer-provider.md

```diff
@@ -0,0 +1,999 @@
+
+## Motivation
+
+In private data center environments, there is no common infrastructure provider for load balancers vs. the provider for the machines themselves. Virtualized environments may use one vendor for virtual machines, e.g. VMware vSphere, KVM, or Xen, and a variety of vendors for networks and load balancers, e.g. F5, VMware NSX-T, Cisco ACI etc… This requires the ability to plug in different infrastructure providers for networking constructs with that of the machine provisioner. The most common use is the provisioning of a load balancer to sit in front of the Kubernetes control plane instances in order to provide a stable endpoint. Many cluster operators will also want to provision a stable machine-based load balancer to front all workload services from that provisioned cluster.
+
+Additionally, there are also cases where providers such as vSphere may be run in public cloud environments and being able to re-use existing native load balancer implementations would be preferred compared to having to re-implement or fork existing implementations for use in those environments.
+
+There are also cases within existing infrastructure provider implementations where it would be beneficial to allow for easier swapping of different load balancer implementations that may provide different benefits under different circumstances.
+
+We've also seen use cases where there is a need for a Load Balancer to front non-control plane services, such as Ingress. Any functionality that we add should also support these use cases in addition to just control plane load balancing.
+
+### Goals
+
+- Provide a common abstraction for Load Balancers between various providers
+- Allow for reuse of load balancer providers between infrastructure providers where possible
+- Load Balancer must be able to interoperate with alternative Control Plane Providers
+- Continued reconciliation of members of the Load Balancer (quiescing of connections, allow for graceful termination/removal of membership as part of deletion mechanisms)
+- Should provide an upgrade path for existing Clusters (specifically vsphere clusters and the current Load Balancer provider)
+- Should not be tied to a specific Control Plane provider implementation (to support alternative Control Plane provider implementations)
+- Should not be a required component of a "Cluster", since some control plane implementations do not require a load balancer (for example, Control Plane providers based on managed Kubernetes services)
+- Must be optional to start, to limit the changes required of infrastructure providers
```

Comment:
Given the complexity that having separate LB CRs adds (moar CRs for users to understand, extra controllers to run and update), do we plan to enforce this on all providers in the long term?

Could we consider this to be an "if you need it" or power user feature and change this requirement to `Must be optional`?